connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -509,7 +509,7 @@ public SourceTaskMetricsGroup(ConnectorTaskId id, ConnectMetrics connectMetrics)
             pollTime.add(metricGroup.metricName(registry.sourceRecordPollBatchTimeMax), new Max());
             pollTime.add(metricGroup.metricName(registry.sourceRecordPollBatchTimeAvg), new Avg());
 
-            sourceRecordActiveCount = metricGroup.metrics().sensor("sink-record-active-count");
+            sourceRecordActiveCount = metricGroup.metrics().sensor("source-record-active-count");
```

Comment:
I think the reported bug might actually the fact that we use `metricGroup.metrics().sensor()` instead of `metricGroup.sensor()`. Since we are not going through the `MetricGroup`, the sensor will not be unique across source tasks. That means when we update this sensor for one source task, we will update all the other source tasks as well. This seems to fit with what was reported in the JIRA. You can verify the bug by adding another metric group to the test case with a different taskId.

Looking over the code, it looks like `WorkerSinkTask` also suffers from the same problem. We should fix that as well and add test cases which verify sensor uniqueness. It would also be a good idea to add tests which verify that metrics and sensors are removed correctly for a group after closing the metric group.

Reply:
You are right, the sensor creation was wrong.  I fixed the set up of sensors and added unit tests that verify that the sensors were not unique before the fix, but unique after the fix.